test(transformer): generate change detectors in template compiler tests

munificent opened the pull request:
This doesn't actually run them, of course, but it at least validates
the transformer output is what we expect.

munificent commented:
cc @jakemac53

jakemac53 commented on modules/angular2/test/transform/template_compiler/duplicate_files/expected/hello.ng_deps.dart:
```diff
@@ -19,4 +22,118 @@ void initReflector(reflector) {
+    }
+    currentProto = _protos[1];
+    greeting1 = context.greeting;
+    if (!_gen.looseIdentical(greeting1, _greeting1)) {
```
Not a blocker for submitting this cl, but this should probably be optimized, `greeting` is being checked twice.

jakemac53 commented on modules/angular2/test/transform/template_compiler/inline_method_files/expected/hello.ng_deps.dart:
```diff
@@ -19,4 +22,70 @@ void initReflector(reflector) {
+  _gen.preGeneratedProtoDetectors['HelloCmp_comp_0'] =
+      _HelloCmp_ChangeDetector0.newProtoChangeDetector;
+}
+class _HelloCmp_ChangeDetector0 extends _gen.AbstractChangeDetector {
```
We should also be creating a much smaller stub class in this case most likely? Dart2js might be doing a decent job at removing dead code here though.

jakemac53 commented:
Left a few comments, they are not blockers for this pr though, more just general comments that should possibly turn into issues.

jakemac53 commented:
@vsavkin or @tbosch, can you review this as well? These are likely to need to be updated somewhat regularly, but its good to know how changes are effecting actual generated code IMO.

munificent committed:
test(transformer): generate change detectors in template compiler tests

This doesn't actually run them, of course, but it at least validates
the transformer output is what we expect.

munificent commented:
@vsavkin or @tbosch, friendly ping for a review please.

tbosch commented on modules/angular2/test/transform/template_compiler/all_tests.dart:
```diff
@@ -12,36 +12,27 @@ import 'package:guinness/guinness.dart';
   // dart browser is in sync. Change this to "not contains notifyBinding"
   // when https://github.com/angular/angular/issues/3019 is solved.
-  it('shouldn always notifyBinding for template variables', () async {
+  it('should not always have notifyBinding for template variables', () async {
```
Right now, we actually do expect a `notifyBinding` always. This was part of a temporary fix to sync the server dom adapter with the browser dom adapter, so they never remove property bindings. The right solution is to use a schema, but this will be done later. so the test description was actually correct.

tbosch commented:
We might want to think about the tests. These tests will all break on every change to the change detector generator, and there is no way whether the changed version is actually correct. 

Why do we need to have the test whether change detector generation works correctly everywhere?

munificent commented:
>  These tests will all break on every change to the change detector generator

True, but I think that's generally true of tests for compilers/transpilers. You test it against the actual expected output code. The change generator transformer basically is a transpiler, so this doesn't seem unreasonable to me.

But if you think this will be too brittle or annoying to maintain when the transformer changes, I'm not personally attached to this. @jakemac53 thought it would be helpful to do, so I did.

vsavkin commented:
I am with Tobias here.

I think a better strategy would be to verify that the dynamic and generated detectors have the same behavior. Currently, we are doing it on the unit test level, but we lack good integration tests that do the same. 

I agree with you. The change detector generator is a transpiler. And there is definitely value in having tests verifying its output (e.g., documentation.) But in my mind, we should have only a couple of those. We already have one such a test. Maybe we should extend it to cover more interesting use cases.

Where do you see the most value of these tests? Troubleshooting or correctness?

munificent commented:
> Where do you see the most value of these tests? Troubleshooting or correctness?

Both. Verying the behavior of the generated code is most important to make sure it actually does what it's supposed to. But there is a lot of ways code can do the right thing but not be ideal. In particular, it can be too large or too slow.

If you're comfortable relying on benchmarks for those, then validating the actual generated code may not be needed. But otherwise, it might be handy to test exactly what kind out of output you get.

This is really up to you folks. If you're happy with your existing coverage, I'm happy to close this.

tbosch commented:
Ok, let's close it. I think what we need are tests that run the compiler with generated change detectors in karma, so they are easy to debug but still cover a lot.

jakemac53 commented:
Sorry for the late response, but here are my thoughts. The purpose of this was threefold:

- Debugging: The best way to debug transformers is with a unit test. Having something already set up that you can hook a debugger into saves a lot of time when debugging.
- Documentation: Design docs are often hard to find or non-existent. Tests like this are very useful for documentation purposes because they will always remain with the code, and provide a good starting point for somebody who is trying to understand what change detectors are and how they work.
- Sanity Checking: When modifying the generator, it is important to understand how it actually affects real generated code. By forcing you to update tests whenever you change the generator, it forces you to evaluate the consequences of what you are doing. It also makes it easier to review those pull requests because you can look at the test to see what the actual expected change is.

That being said, I agree it was probably overkill to add this to all template compiler tests. We should instead have a separate test or suite of tests which verify each feature. I created an issue to track that.
